### ISSUE TITLE
A ShouldThrow method for ArgumentNullException

### ISSUE THREAD

**Issue by josemesona** [shouldly/shouldly#392]
I have some places where I'm testing that a method throws an ArgumentException or ArgumentNullException and has the correct argument name.  I wrote my own method for now to allow me to reuse this functionality but was hoping to add it to Shouldly.

``` C#
protected TException ShouldThrow<TException>(string paramName, Action action)
    where TException : ArgumentException
{
    // Let Shoudly ensure the correct exception type is thrown
    //
    T exception = Should.Throw<T>(action);

    // Verify that the ParamName property is correct since it's known to be an ArgumentException
    //
    exception.ParamName.ShouldBe(paramName);
    
    return exception;
}
```

Can I add this as an overload to ``Should.Throw<T>``?

**Comment by jeffkl** [comment]
I'd like to add this, will you accept a PR?

**Comment by jnm2** [comment]
I like this concept. I like it even more when the message is included, because I've been bitten by tests only checking paramName and not realizing the test only passed because of an unrelated ArgumentException for that parameter due to a factor it wasn't even testing for.

Is it better to add `paramName` alone or should we require a `message` parameter too, to push people in that direction? What do you think?

/cc @shouldly/core

**Comment by jeffkl** [comment]
I think the pattern is usually:

```C#
public void Method(string param1)
{
    if(param1 == null)
    {
      throw new ArgumentNullException(nameof(param1));
    }
}
```

In which case all I've specified is the `paramName`.  We can assume the message will be correct as it comes from the runtime.

But there is [an overload](https://docs.microsoft.com/en-us/dotnet/api/system.argumentnullexception.-ctor?view=net-5.0#System_ArgumentNullException__ctor_System_String_System_String_) to `ArgumentException` that takes a message so there might as well be an overload to a `ShouldThrow` that lets you specify an expected message.

**Comment by jnm2** [comment]
Good point. It makes sense to verify the message of all the ArgumentExceptions except for ArgumentNullException, pretty much, but that's one message you really don't want to be specifying in test code.

I'm happy with either then, just a `string paramName` overload, or both that and a `string paramName, string message` overload.

**Comment by jnm2** [comment]
@jeffkl I'll certainly take you up on your generous offer if you still have the desire after three years :D I'll remove `Up For Grabs` and assign you if you'd like.

**Comment by jeffkl** [comment]
@jnm2 Yes I can contribute this back, please assign it to me.

**Comment by jeffkl** [comment]
I started working on this but hit an issue.  `ShouldThrow` has default parameters so when I added a string argument the compiler is unhappy.

Should I introduce a new method called `ShouldThrowArgumentException<ArgumentException>()`?  Its a little verbose, any suggestions or thoughts?

**Comment by jnm2** [comment]
Can you add the string argument before the default parameters? If I remember right, `string? customMessage = null` should always be the last parameter no matter what; otherwise, reading the call site could be confusing.

**Comment by jnm2** [comment]
We could also consider making this composable, e.g. `.ShouldThrow<ArgumentException>().ShouldHave("argName", "Optional message);`. The downside would be discoverability and not having as much of a strong suggestion that you should verify the param name if it's an argument exception.

**Comment by jeffkl** [comment]
Here's a better example of what I was trying to do:

![image](https://user-images.githubusercontent.com/17556515/116463428-87c1a700-a81f-11eb-8e12-d95f1335f721.png)

This of course works but is slightly verbose:
```C#
public static TException ThrowArgumentException<TException>([InstantHandle] Action actual, string paramName, string? customMessage = null)
    where TException : ArgumentException
{
    TException exception = ThrowInternal<TException>(actual, customMessage: customMessage);

    exception.ParamName.ShouldBe(paramName, customMessage: customMessage);

    return exception;
}
```

The call is laughable:

```C#
Should.ThrowArgumentException<ArgumentException>(() => new Thing(null), "param1");
```
I don't really like the other way of calling it (and its the main point of this being more built-in):
```C#
Should.Throw(() => { new Thing(null) })
  .ParamName.ShouldBe("param1");
```

It works last long as you specify the param name:
```C#
Should.Throw<ArgumentException>(() => new Thing(null), paramName: "param1");
```

But if you don't specify the name, you get the other call:
```C#
// "param1" is actually mapped to customMessage
Should.Throw<ArgumentException>(() => new Thing1(null), "param1");
```

The generic type doesn't behave how I expected, maybe this isn't possible to introduce in a user friendly way.

**Comment by hemiaoio** [comment]
I think it should not be limited to ArgumentException. Should it be suitable for all Exception types, and then you can judge the value of the property in the type, such as the ErrorCode property of HasErrorCodeException
